### PR TITLE
Bug 870169 - Extract reusable telemetry code from ANRReporter.java v2

### PR DESCRIPTION
**Pull request by liuche**
Addressed comments from pull #320 (minus perf concerns about holding onto buffers/strings/etc).

Added a few basic tests, do suggest more if needed.

### CODE REVIEW

**Comment by rnewman** [comment]
The reason is almost certainly to avoid having to parse that JSON when the slug is uploaded.

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+ *
+ * A saved telemetry ping file consists of JSON in the following format,
+ *   {
+ *     "slug": "<uuid-string>",
+ *     "payload": "<escaped-json-data-string>",
+ *     "checksum": "<base64-sha-256-string>"
+ *   }
+ *
+ * This class writes first to a temporary file and then, after finishing the contents of the ping,
+ * copies that to the file specified by the caller. This is to avoid uploads of partially written
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:35]: Moves, not copies, right?

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+ * ping files.
+ *
+ * The API provided by this class:
+ * startPingFile() - opens stream to a tmp File and writes the slug
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:39]: Writes the slug _header_, no?

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+ * appendPayload(String payloadContent) - appends to the payload of the ping and updates the
+ *                                        checksum
+ * finishPingFile() - writes the checksum to the tmp file and copies it to the File specified by the caller.
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:42]: Moves.

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+   * @throws Exception
+   *           Checked exceptions <code>NoSuchAlgorithmException</code>,
+   *           <code>UnsupportedEncodingException</code>, or
+   *           <code>IOException</code> and unchecked exception that
+   *           are rethrown to caller
+   */
+  public void startPingFile() throws Exception {
+
+    // Open stream to temporary file for writing.
+    String tmpFilename = filename + ".tmp";
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:122]: Nope!

http://stackoverflow.com/questions/3425906/creating-temporary-files-in-android

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    tmpFile = new File(parentDir, tmpFilename);
+    if (tmpFile.exists()) {
+        tmpFile.delete();
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:125]: Indenting.

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    File destFile = new File(parentDir, filename);
+    // Delete file if it exists - docs state that rename may fail if the File already exists.
+    if (destFile.exists()) {
+      destFile.delete();
+    }
+    boolean result = tmpFile.renameTo(destFile);
+    if (!result) {
+      throw new IOException("Could not move tmp file to destination.");
+    }
+    cleanUp();
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:239]: What does `cleanup()` do to `tmpFile`?

I suspect the answer is "it's fine, `tmpFile.exists()` returns false so it doesn't delete", but let's go ahead and null-out `tmpFile` before calling `cleanUp`, if you even care at all about discarding the checksum.

Also, shouldn't this be in a `finally` block around the whole last block?

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+  private void cleanUp() {
+    // Discard checksum.
+    checksum.reset();
+
+    // Clean up files.
+    if (tmpFile != null && tmpFile.exists()) {
+      tmpFile.delete();
+    }
+  }
+  /**
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:274]: Newline before.

File: test/src/org/mozilla/gecko/background/datareporting/test/TestTelemetryRecorder.java [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+import android.util.Base64;
+
+public class TestTelemetryRecorder extends FakeProfileTestCase {
+  private File telemetryPingDir;
+  private TelemetryRecorder telemetryRecorder;
+
+  private static final String DEST_FILENAME = "dest-filename";
+  private final String TEST_PAYLOAD1 = "{\"ver\": 1, \"measurements\":" +
+		                                 "{ \"uptime\": 24982 }, \"data\": {";
```
Comment (rnewman) [inline test/src/org/mozilla/gecko/background/datareporting/test/TestTelemetryRecorder.java:31]: Indenting.

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+      Logger.debug(LOG_TAG, "Wrote " + footer.length + " footer bytes.");
+    } catch (UnsupportedEncodingException e) {
+      cleanUpAndRethrow("Checksum encoding exception", e);
+    } catch (IOException e) {
+      cleanUpAndRethrow("Error writing footer to stream", e);
+    } finally {
+      try {
+        outputStream.close();
+      } catch (IOException e) {
+        // Failed to close, nothing we can do.
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:225]: Is this a situation you just want to quietly ignore? You want to go ahead with the move?

**Comment by liuche** [comment]
Closing and updating to pull #326.

File: src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java (hunk 9 above)
Comment (liuche) [inline src/main/java/org/mozilla/gecko/background/datareporting/TelemetryRecorder.java:225]: I think so? The contents have been flushed, so I will null out the reference to the stream so nothing else can be written to it. There's really nothing else we can do, right?

**Pull request closed without merging**